File: Unity/WebSocketApp/Assets/Scripts/SocketMessage.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.IO.Compression;
using UnityEngine;

namespace WebSocketApp
{
    public enum UnityDirectoryType
    {
        Invalid = -1,

        Data,
        StreamingAssets,
        PersistentData,
        TemporaryCache,
    } // enum UnityDirectoryType

    //---------------------------------

    [Serializable]
    public class SocketMessageBase
    {
        private static readonly Dictionary<string, Type> MESSAGE_TYPES = new Dictionary<string, Type>()
        {
            {typeof(SockeTextMessage).Name,  typeof(SockeTextMessage)},
            {typeof(SocketRequestMessage).Name,  typeof(SocketRequestMessage)},
            {typeof(SocketConnectionInformationMessage).Name,  typeof(SocketConnectionInformationMessage)},
            {typeof(SocketScreenShotRequestMessage).Name,  typeof(SocketScreenShotRequestMessage)},
            {typeof(SocketFileListRequestMessage).Name,  typeof(SocketFileListRequestMessage)},
            {typeof(SocketFileUploadRequestMessage).Name,  typeof(SocketFileUploadRequestMessage)},
            {typeof(SocketConnectGameObjectRequestMessage).Name,  typeof(SocketConnectGameObjectRequestMessage)},
            {typeof(SocketFileMessage).Name,  typeof(SocketFileMessage)},
            {typeof(SocketScreenShotMessage).Name,  typeof(SocketScreenShotMessage)},
            {typeof(SocketMoveGameObjectMessage).Name,  typeof(SocketMoveGameObjectMessage)},
            {typeof(SocketImageDataMessage).Name,  typeof(SocketImageDataMessage)},
        };

        public string MessageType
        {
            get => _messageType;
        }
        public string _messageType;

        private static byte[] _base64Buffer = new byte[1024];


        public SocketMessageBase()
        {
            _messageType = GetType().Name;
        }

        public static SocketMessageBase ImportMessage(string val)
        {
            var index = val.IndexOf(',');
            if (index < 0)
            {
                Debug.LogErrorFormat("メッセージの書式が不正：{0}", val);
                return null;
            }

            var typeKey = val.Substring(0, index);
            if (!MESSAGE_TYPES.TryGetValue(typeKey, out var type))
            {
                Debug.LogErrorFormat("未対応のメッセージタイプ：{0}", typeKey);
                return null;
            }
            ++index;

            bool compressed = (val[index] == 'c');
            index += 2;

            SocketMessageBase message = null;
            try
            {
                var json = DecodeBase64(val.Substring(index), compressed);
                message = JsonUtility.FromJson(json, type) as SocketMessageBase;
            }
            catch (Exception ex)
            {
                Debug.LogException(ex);
                return null;
            }

            return message;
        }

        public static string ExportMessage(SocketMessageBase message)
        {
            string result = message.MessageType + ",";

            try
            {
                var json = JsonUtility.ToJson(message);

                if (string.IsNullOrEmpty(json))
                {
                    result += "-,";
                } else
                {
                    var length = System.Text.Encoding.UTF8.GetByteCount(json);
                    var jsonBuffer = GetEncodeBuffer(length);
                    var jsonBufferLength = System.Text.Encoding.UTF8.GetBytes(json, 0, json.Length, jsonBuffer, 0);
                    var compressedArray = Compress(jsonBuffer, 0, jsonBufferLength);
                    if (compressedArray == null || compressedArray.Length >= jsonBufferLength)
                    {
                        result += "-,";
                        result += Convert.ToBase64String(jsonBuffer, 0, jsonBufferLength);
                    }
                    else
                    {
                        result += "c,";
                        result += Convert.ToBase64String(compressedArray, 0, compressedArray.Length);
                    }
                }
            }
            catch(Exception ex)
            {
                Debug.LogException(ex);
                return "";
            }

            return result;
        }

        public static string GetDirectory(UnityDirectoryType directoryType)
        {
            switch (directoryType)
            {
                case UnityDirectoryType.Data:
                    return Application.dataPath;
                case UnityDirectoryType.StreamingAssets:
                    return Application.streamingAssetsPath;
                case UnityDirectoryType.PersistentData:
                    return Application.persistentDataPath;
                case UnityDirectoryType.TemporaryCache:
                    return Application.temporaryCachePath;

                default:
                    break;
            }
            return "";
        }

        public static byte[] Compress(byte[] src, int srcOffset, int srcLength)
        {
            using (MemoryStream outMemoryStream = new MemoryStream())
            {
                using (GZipStream gzipStream = new GZipStream(outMemoryStream, System.IO.Compression.CompressionLevel.Fastest))
                {
                    gzipStream.Write(src, srcOffset, srcLength);
                }
                return outMemoryStream.ToArray();
            }
        }

        public static byte[] Decompress(byte[] bytes)
        {
            using (MemoryStream inStream = new MemoryStream(bytes))
            {
                using (MemoryStream outStream = Decompress(inStream))
                {
                    if (outStream != null)
                    {
                        return outStream.ToArray();
                    }
                }
            }
            return null;
        }

        public static MemoryStream Decompress(Stream inStream)
        {
            MemoryStream outStream = new MemoryStream(capacity: (int)inStream.Length);
            using (GZipStream gzipStream = new GZipStream(inStream, CompressionMode.Decompress))
            {
                int length = 1024;
                byte[] buffer = new byte[length];
                while ((length = gzipStream.Read(buffer, 0, buffer.Length)) > 0)
                {
                    outStream.Write(buffer, 0, length);
                }
                outStream.Position = 0;
            }
            return outStream;
        }
        public static string EncodeBase64(string src, out bool compressed)
        {
            compressed = false;
            if (string.IsNullOrEmpty(src))
            {
                return "";
            }

            var length = System.Text.Encoding.UTF8.GetByteCount(src);
            var buffer = GetEncodeBuffer(length);
            var bufferLength = System.Text.Encoding.UTF8.GetBytes(src, 0, src.Length, buffer, 0);
            return Convert.ToBase64String(buffer, 0, bufferLength);
        }

        public static string DecodeBase64(string base64Encoded, bool compressed = false)
        {
            char[] encodedArray = base64Encoded.ToCharArray();
            var decodedArray = Convert.FromBase64CharArray(encodedArray, 0, encodedArray.Length);
            if (!compressed)
            {
                return System.Text.Encoding.UTF8.GetString(decodedArray);
            }
            else
            {
                var compressedArray = Decompress(decodedArray);
                if (compressedArray == null)
                {
                    return "";
                }
                return System.Text.Encoding.UTF8.GetString(compressedArray);
            }
        }

        private static byte[] GetEncodeBuffer(int length)
        {
            return (length <= _base64Buffer.Length ? _base64Buffer : (_base64Buffer = new byte[length]));
        }
    } // class SocketMessageBase

    //---------------------------------

    [Serializable]
    public class SockeTextMessage : SocketMessageBase
    {
        public static readonly string MESSAGE_TYPE = typeof(SockeTextMessage).Name;

        public string _text;
    } // class SockeTextMessage

    //---------------------------------

    [Serializable]
    public class SocketRequestMessage : SocketMessageBase
    {
        public static readonly string MESSAGE_TYPE = typeof(SocketRequestMessage).Name;

        public string _request;
        public int _requestId;
    } // class SocketRequestMessage

    //---------------------------------

    [Serializable]
    public class SocketScreenShotRequestMessage : SocketRequestMessage
    {
        public static readonly new string MESSAGE_TYPE = typeof(SocketScreenShotRequestMessage).Name;

        public bool _stop = false;
        public float _interval = -1f;
    } // class SocketScreenShotRequestMessage

    //---------------------------------

    [Serializable]
    public class SocketFileListRequestMessage : SocketRequestMessage
    {
        public static readonly new string MESSAGE_TYPE = typeof(SocketFileListRequestMessage).Name;

        public UnityDirectoryType _directoryType = UnityDirectoryType.Invalid;
        public string _targetPath;
    } // class SocketFileListRequestMessage

    //---------------------------------

    [Serializable]
    public class SocketFileUploadRequestMessage : SocketRequestMessage
    {
        public static readonly new string MESSAGE_TYPE = typeof(SocketFileUploadRequestMessage).Name;

        public UnityDirectoryType _directoryType = UnityDirectoryType.Invalid;
        public string _targetPath;
    } // class SocketFileUploadRequestMessage

    //---------------------------------

    [Serializable]
    public class SocketConnectGameObjectRequestMessage : SocketRequestMessage
    {
        public static readonly new string MESSAGE_TYPE = typeof(SocketConnectGameObjectRequestMessage).Name;

        public string _gameObjectName;
    } // class SocketConnectGameObjectRequestMessage

    //---------------------------------

    [Serializable]
    public class SocketLogMessage : SocketMessageBase
    {
        public static readonly string MESSAGE_TYPE = typeof(SocketLogMessage).Name;

        public SocketLogMessage(LogType logType, string log, string stackTrace)
        {
            var now = DateTime.Now;
            _logType = logType;
            _time = string.Format("{0:00}:{1:00}:{2:00}", now.Hour, now.Minute, now.Second);
            _log = log;
            _stackTrace = stackTrace;
        }

        public LogType _logType;
        public string _time;
        public string _log;
        public string _stackTrace;
    }; // class SocketLogMessage

    //---------------------------------

    [Serializable]
    public class SocketConnectionInformationMessage : SocketMessageBase
    {
        public static readonly string MESSAGE_TYPE = typeof(SocketConnectionInformationMessage).Name;

        public SocketConnectionInformationMessage(int requestId) : base()
        {
            _requestId = requestId;

            try
            {
                _applicationName = Application.productName;
                _uuid = SystemInfo.deviceUniqueIdentifier;
                _deviceName = UnityEngine.SystemInfo.deviceName;
#if UNITY_EDITOR
                _deviceModel = UnityEngine.SystemInfo.deviceModel;
#elif UNITY_IPHONE
                _deviceModel = UnityEngine.iOS.Device.generation.ToString();
#elif UNITY_ANDROID
                _deviceModel = UnityEngine.SystemInfo.deviceModel;
#endif
            }
            catch (Exception ex)
            {
                Debug.LogException(ex);
            }
        }

        public int _requestId;
        public string _applicationName;
        public string _uuid;
        public string _deviceName;
        public string _deviceModel;
    } // class SocketConnectionInformationMessage

    //---------------------------------

    [Serializable]
    public class SocketFileListMessage : SocketMessageBase
    {
        public static readonly string MESSAGE_TYPE = typeof(SocketFileListMessage).Name;

        public SocketFileListMessage(int requestId) : base()
        {
            _requestId = requestId;
        }

        public void UpdateList(UnityDirectoryType directoryType, string targetPath)
        {
            _files = null;
            _directories = null;

            _directory = GetDirectory(directoryType);
            var path = (string.IsNullOrEmpty(targetPath) ? _directory : Path.Combine(_directory, targetPath));
            if (!Directory.Exists(path))
            {
                return;
            }

            _files = Directory.GetFiles(path);
            _directories = Directory.GetDirectories(path);
        }

        public int _requestId;
        public string _directory;
        public string[] _files = null;
        public string[] _directories = null;
    } // class SocketFileListMessage

    //---------------------------------

    [Serializable]
    public class SocketFileMessage : SocketMessageBase
    {
        public static readonly string MESSAGE_TYPE = typeof(SocketFileMessage).Name;

        public SocketFileMessage(int requestId) : base()
        {
            _requestId = requestId;
        }

        public bool SetFile(UnityDirectoryType directoryType, string targetPath)
        {
            _directoryType = directoryType;
            _targetPath = targetPath;

            var directory = GetDirectory(directoryType);
            var path = (string.IsNullOrEmpty(targetPath) ? directory : Path.Combine(directory, targetPath));
            if (!File.Exists(path))
            {
                Debug.LogErrorFormat("指定されたファイルが存在しない：{0}", path);
                return false;
            }

            var bytes = System.IO.File.ReadAllBytes(path);
            if (bytes == null)
            {
                Debug.LogErrorFormat("ファイルの読み込み失敗：{0}", path);
                return false;
            }
            _data = Convert.ToBase64String(bytes);

            return true;
        }

        public byte[] GetFileData()
        {
            if (string.IsNullOrEmpty(_data))
            {
                return null;
            }

            char[] encodedArray = _data.ToCharArray();
            var file = Convert.FromBase64CharArray(encodedArray, 0, encodedArray.Length);
            return file;
        }

        public bool WriteFile()
        {
            var directory = GetDirectory(_directoryType);
            if (!Directory.Exists(directory))
            {
                Debug.LogErrorFormat("ディレクトリが存在しない： {0}", directory);
                return false;
            }

            string fileName = GetFileName();
            string path = Path.Combine(directory, fileName);

            var fileData = GetFileData();
            if (fileData == null)
            {
                return false;
            }

            using (var stream = new FileStream(path, FileMode.Create))
            {
                using (var writer = new BinaryWriter(stream))
                {
                    writer.Write(fileData);
                }
                stream.Close();

                if (!File.Exists(path))
                {
                    Debug.LogErrorFormat("ファイルの書き込みに失敗： {0}", path);
                    return false;
                }
            }

            return true;
        }

        public string GetFileName()
        {
            var index = _targetPath.LastIndexOf('/');
            if (index == -1)
            {
                return _targetPath;
            }
            else
            {
                return _targetPath.Substring(index + 1);
            }
        }

        public UnityDirectoryType _directoryType = UnityDirectoryType.Invalid;
        public string _targetPath;
        public int _requestId;
        public string _data;
    } // class SocketFileMessage

    //---------------------------------

    [Serializable]
    public class SocketImageDataMessage : SocketMessageBase
    {
        public static readonly string MESSAGE_TYPE = typeof(SocketImageDataMessage).Name;

        public byte[] GetImage()
        {
            if (string.IsNullOrEmpty(_imageData))
            {
                return null;
            }

            char[] encodedArray = _imageData.ToCharArray();
            var image = Convert.FromBase64CharArray(encodedArray, 0, encodedArray.Length);
            return image;
        }

        public bool SetImage(byte[] bytes)
        {
            if (bytes == null)
            {
                return false;
            }
            _imageData = Convert.ToBase64String(bytes);

            return true;
        }

        public string _imageData;
    } // class SocketImageDataMessage

    //---------------------------------

    [Serializable]
    public class SocketScreenShotMessage : SocketImageDataMessage
    {
        public new static readonly string MESSAGE_TYPE = typeof(SocketScreenShotMessage).Name;

        public SocketScreenShotMessage(int requestId) : base()
        {
            _requestId = requestId;
        }

        public int _requestId;
        public string _dateTime;
    } // class SocketScreenShotMessage

    //---------------------------------

    [Serializable]
    public class SocketMoveGameObjectMessage : SocketImageDataMessage
    {
        public new static readonly string MESSAGE_TYPE = typeof(SocketMoveGameObjectMessage).Name;

        public SocketMoveGameObjectMessage() : base()
        {
        }

        public int _manipulateTarget;
        public float _x;
        public float _y;
        public float _z;
    } // class SocketMoveGameObjectMessage
} // namespace WebSocketApp

File: Qt/WebSocketApp/SocketMessage.cpp
#include "SocketMessage.h"

const char* SockeTextMessage::MESSAGE_TYPE = "SockeTextMessage";
const char* SocketConnectionInformationMessage::MESSAGE_TYPE = "SocketConnectionInformationMessage";
const char* SocketMoveGameObjectMessage::MESSAGE_TYPE = "SocketMoveGameObjectMessage";
const char* SocketRequestMessage::MESSAGE_TYPE = "SocketRequestMessage";
const char* SocketResponseMessage::MESSAGE_TYPE = "SocketResponseMessage";
const char* SocketScreenShotRequestMessage::MESSAGE_TYPE = "SocketScreenShotRequestMessage";
const char* SocketFileListRequestMessage::MESSAGE_TYPE = "SocketFileListRequestMessage";
const char* SocketFileUploadRequestMessage::MESSAGE_TYPE = "SocketFileUploadRequestMessage";
const char* SocketConnectGameObjectRequestMessage::MESSAGE_TYPE = "SocketConnectGameObjectRequestMessage";
const char* SocketLogMessage::MESSAGE_TYPE = "SocketLogMessage";
const char* SocketFileListMessage::MESSAGE_TYPE = "SocketFileListMessage";
const char* SocketFileMessage::MESSAGE_TYPE = "SocketFileMessage";
const char* SocketImageDataMessage::MESSAGE_TYPE = "SocketImageDataMessage";
const char* SocketScreenShotMessage::MESSAGE_TYPE = "SocketScreenShotMessage";


SocketMessageBase* SocketMessageBase::ImportMessage(const QString& val) {
    auto index = val.indexOf(',');
    if (index == -1) {
        return nullptr;
    }

    const auto& typeKey = val.mid(0, index);
    if (index+1 >= val.length()) {
        return nullptr;
    }
    ++index;

    bool compressed = (val[index] == 'c');
    index += 2;

    QByteArray decodedArray;
    {
        const auto& base64Encoded = val.mid(index);
        if (!compressed) {
            decodedArray = QByteArray::fromBase64(base64Encoded.toUtf8());
        } else {
            auto compressedArray = QByteArray::fromBase64(base64Encoded.toUtf8());

            std::vector<char> decompressed;
            auto decompressedLength = WebSocketApp::DecompressGZip(compressedArray.data(), compressedArray.length(), decompressed);
            if (decompressedLength == -1) {
                OUTPUT_ERROR_LOG("%sの解凍に失敗", typeKey.toUtf8().data());
                return nullptr;
            }
            decodedArray = QByteArray(&(decompressed[0]), decompressed.size());
        }
    }
    QJsonDocument document = QJsonDocument::fromJson(decodedArray);

    SocketMessageBase* message = nullptr;
    {
        if (typeKey == SocketLogMessage::MESSAGE_TYPE) {
            message = new SocketLogMessage();
        } else if (typeKey == SocketConnectionInformationMessage::MESSAGE_TYPE) {
            message = new SocketConnectionInformationMessage();
        } else if (typeKey == SocketFileListMessage::MESSAGE_TYPE) {
            message = new SocketFileListMessage();
        } else if (typeKey == SocketFileMessage::MESSAGE_TYPE) {
            message = new SocketFileMessage();
        } else if (typeKey == SocketScreenShotMessage::MESSAGE_TYPE) {
            message = new SocketScreenShotMessage();
        }
    }
    if (message == nullptr) {
        return nullptr;
    }

    auto obj = document.object();
    if (!message->FromJson(obj)) {
        delete message;
        return nullptr;
    }

    return message;
}

bool SocketMessageBase::ExportMessage(QString& message) const {
    QJsonObject obj;
    if (!ToJson(obj)) {
        return false;
    }

    message = QFORMAT_STR("%s,", _messageType.c_str());

    if (!obj.empty()) {
        QJsonDocument document(obj);

        QByteArray array = document.toJson(QJsonDocument::Compact);
        {
            std::vector<char> compressed;
            if (WebSocketApp::CompressGZip(array.data(), array.length(), compressed) <= 0) {
                message += "-,";
            } else if ((int)compressed.size() >= array.length()) {
                message += "-,";
            } else {
                message += "c,";
                array = QByteArray(&(compressed[0]), compressed.size());
            }
        }
        message += array.toBase64();
    }
    return true;
}

bool SocketMessageBase::ToJson(QJsonObject& obj) const {
    SET_JSON_VALUE(_messageType, obj);
    return true;
}

bool SocketMessageBase::FromJson(QJsonObject& obj) {
    if (!GET_JSON_VALUE(_messageType, obj)) {
        return false;
    }
    return true;
}

//---------------------------------

bool SockeTextMessage::ToJson(QJsonObject& obj) const {
    if (!SocketMessageBase::ToJson(obj)) {
        return false;;
    }

    SET_JSON_VALUE(_text, obj);

    return true;
}

//---------------------------------

int SocketRequestMessage::_nextRequestId = 0;

bool SocketRequestMessage::ToJson(QJsonObject& obj) const {
    if (!SocketMessageBase::ToJson(obj)) {
        return false;;
    }

    SET_JSON_VALUE(_request, obj);
    SET_JSON_VALUE(_requestId, obj);

    return true;
}

//---------------------------------

bool SocketScreenShotRequestMessage::ToJson(QJsonObject& obj) const {
    if (!SocketRequestMessage::ToJson(obj)) {
        return false;;
    }

    SET_JSON_VALUE(_stop, obj);
    SET_JSON_VALUE(_interval, obj);
    return true;
}

//---------------------------------

bool SocketFileListRequestMessage::ToJson(QJsonObject& obj) const {
    if (!SocketRequestMessage::ToJson(obj)) {
        return false;;
    }

    SET_JSON_INT_VALUE(_directoryType, obj);
    SET_JSON_VALUE(_targetPath, obj);
    return true;
}

//---------------------------------

bool SocketFileUploadRequestMessage::ToJson(QJsonObject& obj) const {
    if (!SocketRequestMessage::ToJson(obj)) {
        return false;;
    }

    SET_JSON_INT_VALUE(_directoryType, obj);
    SET_JSON_VALUE(_targetPath, obj);
    return true;
}

//---------------------------------

bool SocketConnectGameObjectRequestMessage::ToJson(QJsonObject& obj) const {
    if (!SocketRequestMessage::ToJson(obj)) {
        return false;;
    }

    SET_JSON_VALUE(_gameObjectName, obj);
    return true;
}

//---------------------------------

bool SocketResponseMessage::ToJson(QJsonObject& obj) const {
    if (!SocketMessageBase::ToJson(obj)) {
        return false;;
    }

    SET_JSON_VALUE(_request, obj);
    SET_JSON_VALUE(_requestId, obj);

    return true;
}

bool SocketResponseMessage::FromJson(QJsonObject& obj) {
    if (!SocketMessageBase::FromJson(obj)) {
        return false;;
    }

    if (!GET_JSON_VALUE(_request, obj)) {
        return false;
    }
    if (!GET_JSON_VALUE(_requestId, obj)) {
        return false;
    }

    return true;
}

//---------------------------------

bool SocketLogMessage::FromJson(QJsonObject& obj) {
    if (!SocketMessageBase::FromJson(obj)) {
        return false;;
    }

    if (!GET_JSON_INT_VALUE(LogType, _logType, obj)) {
        return false;
    }
    if (!GET_JSON_VALUE(_time, obj)) {
        return false;
    }
    if (!GET_JSON_VALUE(_log, obj)) {
        return false;
    }
    if (!GET_JSON_VALUE(_stackTrace, obj)) {
        return false;
    }

    return true;
}

bool SocketLogMessage::ToJson(QJsonObject& obj) const {
    if (!SocketMessageBase::ToJson(obj)) {
        return false;;
    }

    SET_JSON_INT_VALUE(_logType, obj);
    SET_JSON_VALUE(_log, obj);
    SET_JSON_VALUE(_stackTrace, obj);
    return true;
}

//---------------------------------

bool SocketConnectionInformationMessage::FromJson(QJsonObject& obj) {
    if (!SocketMessageBase::FromJson(obj)) {
        return false;;
    }

    if (!GET_JSON_VALUE(_requestId, obj)) {
        return false;
    }
    if (!GET_JSON_VALUE(_applicationName, obj)) {
        return false;
    }
    if (!GET_JSON_VALUE(_uuid, obj)) {
        return false;
    }
    if (!GET_JSON_VALUE(_deviceName, obj)) {
        return false;
    }
    if (!GET_JSON_VALUE(_deviceModel, obj)) {
        return false;
    }

    return true;
}

bool SocketConnectionInformationMessage::ToJson(QJsonObject& obj) const {
    if (!SocketMessageBase::ToJson(obj)) {
        return false;;
    }

    SET_JSON_VALUE(_applicationName, obj);
    SET_JSON_VALUE(_uuid, obj);
    SET_JSON_VALUE(_deviceName, obj);
    SET_JSON_VALUE(_deviceModel, obj);
    return true;
}

//---------------------------------

bool SocketFileListMessage::FromJson(QJsonObject& obj) {
    if (!SocketMessageBase::FromJson(obj)) {
        return false;;
    }

    if (!GET_JSON_VALUE(_requestId, obj)) {
        return false;
    }
    if (!GET_JSON_VALUE(_directory, obj)) {
        return false;
    }
    if (!ImportFiles(obj)) {
        return false;;
    }
    if (!ImportDirectories(obj)) {
        return false;;
    }

    return true;
}

bool SocketFileListMessage::ImportFiles(QJsonObject& obj) {
    _files.clear();

    QJsonValueRef value = obj["_files"];
    if (!value.isArray()) {
        return false;
    }

    auto files = value.toArray();
    for (int i = 0, count = files.count(); i < count; ++i) {
        _files.push_back(files[i].toString().toUtf8().data());
    }

    return true;
}

bool SocketFileListMessage::ImportDirectories(QJsonObject& obj) {
    _directories.clear();

    QJsonValueRef value = obj["_directories"];
    if (!value.isArray()) {
        return false;
    }

    auto directories = value.toArray();
    for (int i = 0, count = directories.count(); i < count; ++i) {
        _directories.push_back(directories[i].toString().toUtf8().data());
    }

    return true;
}

//---------------------------------

bool SocketImageDataMessage::SetImage(const std::string& imagePath) {
    std::vector<char> buffer;
    if (!WebSocketApp::ReadFile(imagePath, buffer)) {
        return false;
    }

    QByteArray array(&(buffer[0]), buffer.size());
    _imageData = array.toBase64().data();

    return true;
}

bool SocketImageDataMessage::FromJson(QJsonObject& obj) {
    if (!SocketMessageBase::FromJson(obj)) {
        return false;;
    }

    if (!GET_JSON_VALUE(_imageData, obj)) {
        return false;
    }
    _bytes = QByteArray::fromBase64(QByteArray(_imageData.c_str(), _imageData.size()));

    return true;
}

bool SocketImageDataMessage::ToJson(QJsonObject& obj) const {
    if (!SocketMessageBase::ToJson(obj)) {
        return false;;
    }

    SET_JSON_VALUE(_imageData, obj);
    return true;
}

//---------------------------------

bool SocketFileMessage::SetFile(const std::string& dataPath, UnityDirectoryType directoryType /*= UnityDirectoryType::Invalid*/) {
    std::vector<char> buffer;
    if (!WebSocketApp::ReadFile(dataPath, buffer)) {
        return false;
    }

    QByteArray array(&(buffer[0]), buffer.size());
    _data = array.toBase64().data();

    _directoryType = directoryType;
    _targetPath = dataPath;

    return true;
}

bool SocketFileMessage::FromJson(QJsonObject& obj) {
    if (!SocketMessageBase::FromJson(obj)) {
        return false;;
    }

    if (!GET_JSON_VALUE(_requestId, obj)) {
        return false;
    }
    if (!GET_JSON_INT_VALUE(UnityDirectoryType, _directoryType, obj)) {
        return false;
    }
    if (!GET_JSON_VALUE(_targetPath, obj)) {
        return false;
    }

    if (!GET_JSON_VALUE(_data, obj)) {
        return false;
    }
    _bytes = QByteArray::fromBase64(QByteArray(_data.c_str(), _data.size()));

    return true;
}

bool SocketFileMessage::ToJson(QJsonObject& obj) const {
    if (!SocketMessageBase::ToJson(obj)) {
        return false;;
    }

    SET_JSON_VALUE(_requestId, obj);
    SET_JSON_INT_VALUE(_directoryType, obj);
    SET_JSON_VALUE(_targetPath, obj);
    SET_JSON_VALUE(_data, obj);
    return true;
}

//---------------------------------

bool SocketScreenShotMessage::FromJson(QJsonObject& obj) {
    if (!SocketImageDataMessage::FromJson(obj)) {
        return false;;
    }

    if (!GET_JSON_VALUE(_requestId, obj)) {
        return false;
    }
    if (!GET_JSON_VALUE(_dateTime, obj)) {
        return false;
    }

    return true;
}

//---------------------------------

bool SocketMoveGameObjectMessage::FromJson(QJsonObject& obj) {
    if (!SocketMessageBase::FromJson(obj)) {
        return false;;
    }

    if (!GET_JSON_VALUE(_manipulateTarget, obj)) {
        return false;
    }
    if (!GET_JSON_VALUE(_x, obj)) {
        return false;
    }
    if (!GET_JSON_VALUE(_y, obj)) {
        return false;
    }
    if (!GET_JSON_VALUE(_z, obj)) {
        return false;
    }

    return true;
}

bool SocketMoveGameObjectMessage::ToJson(QJsonObject& obj) const {
    if (!SocketMessageBase::ToJson(obj)) {
        return false;;
    }

    SET_JSON_VALUE(_manipulateTarget, obj);
    SET_JSON_VALUE(_x, obj);
    SET_JSON_VALUE(_y, obj);
    SET_JSON_VALUE(_z, obj);
    return true;
}
